### PR TITLE
[CODEGEТ-1918] Copy RESTfulComponent from mlpiper to allow support configuration of uwsig threads

### PR DESCRIPTION
**Pull request by yura-ant**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Copy  RESTfulComponent, UwsgiBroker, WSGI_INI_CONTENT from mlpiper and use it version instead of mlpiper versions:
https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful_component.py
https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful/uwsgi_broker.py
https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful/uwsgi_ini_template.py

Functionality is not changed but we have ability to modify them now to allow improve configuration of uWSGI in separate PR

### CODE REVIEW

**Comment by yakov-g** [comment]
So this is just copying files into this repo, right? DRUM will not use files yet?

**Comment by yakov-g** [comment]
Could you please push into the branch again to trigger tests.

**Comment by yakov-g** [comment]
> So this is just copying files into this repo, right? DRUM will not use files yet?

We should completely remove: mlpiper usage from `uwsgi_serving.py` and remove `component.json` , which is component configuration for mlpiper. And also create a boilerplate to instantiate comopnents without mlpiper.

**Comment by yura-ant** [comment]
> So this is just copying files into this repo, right? DRUM will not use files yet?

There are two commits in this PR. The first one just copy file and not use them. The second one use them via modify import to those copied files. I expect all test should pass as logic of those files are not changed.

**Comment by yura-ant** [comment]
> > So this is just copying files into this repo, right? DRUM will not use files yet?
> 
> We should completely remove: mlpiper usage from `uwsgi_serving.py` and remove `component.json` , which is component configuration for mlpiper. And also create a boilerplate to instantiate comopnents without mlpiper.

In scope of https://docs.google.com/document/d/1yq-T7-pmjW49RsvbslNIl3BF_GRbYLti2McVmZgd_To/edit#heading=h.8vln2nt42rgh I have no plans to remove mlpliper. I've added this as possible future work. I like idea to remove it but I afraid it could lead to much more changes as uwsgi_serving.py is not  just  class which use mlpiper. Also I copied classes where I plan to do changes to allow configure uWSGI threads. But those files as well have deps to  other mlpipers files which I not copied.  In scope of  https://datarobot.atlassian.net/browse/PBMP-6561 I would like to do minimum acceptable changes  to support configure threads

**Comment by yura-ant** [comment]
> Could you please push into the branch again to trigger tests.
just bump version

**Comment by yakov-g** [comment]
In mlpiper in this folder: https://github.com/datarobot/mlpiper/tree/master/mlpiper-py/mlpiper/components - we have components of the **_mlpiper pipeline_**, one of them is RESTfulComponent (restful_component.py )
In DRUM we have UwsgiComponent(RESTfulComponent) https://github.com/datarobot/datarobot-user-models/blob/master/custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/uwsgi_serving.py - in just implements RESTfulComponent.

Back to mlpiper: 
In https://github.com/datarobot/mlpiper/tree/master/mlpiper-py/mlpiper/components/restful - we have whole orthogonal implementation of uwsgi/nginx not related to mlpiper at all (it uses some utils/exceptions, should be changed to DRUM exception, or even its own). (It just can be a separate project).

We should move all the https://github.com/datarobot/mlpiper/tree/master/mlpiper-py/mlpiper/components/restful somewhere into DRUM and keep it a separate  **uwsgi/nginx** implementation. For example to: https://github.com/datarobot/datarobot-user-models/tree/master/custom_model_runner/datarobot_drum/resource/components/restful.. Or even on higher levels.

As for the RESTfulComponent https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful_component.py, we should move it next to UwsgiComponent(RESTfulComponent) https://github.com/datarobot/datarobot-user-models/blob/master/custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/uwsgi_serving.py  (as you did). I would clean it and merge together.

In this way: we will use mlpiper's pipleline construction functionality and component's functionality(nginx) will be in DRUM

I don't think we should take some files from the standalone uwsgi/nginx implementation
https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful_component.py
https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful/uwsgi_broker.py
https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful/uwsgi_ini_template.py
and copy them directly into  `uwsgi_component/uwsgi_serving.py`

In this case we use  mlpiper's pipleline construction functionality and in pararralel use some uswgi/nginx functionality. Which is not good imho.

@zohar-mizrahi ?

**Comment by yakov-g** [comment]
Also, somehow tests don't like your PR, please rebase it on master.

**Comment by yura-ant** [comment]
> In mlpiper in this folder: https://github.com/datarobot/mlpiper/tree/master/mlpiper-py/mlpiper/components - we have components of the **_mlpiper pipeline_**, one of them is RESTfulComponent (restful_component.py ) In DRUM we have UwsgiComponent(RESTfulComponent) https://github.com/datarobot/datarobot-user-models/blob/master/custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/uwsgi_serving.py - it just implements RESTfulComponent.
> 
> Back to mlpiper: In https://github.com/datarobot/mlpiper/tree/master/mlpiper-py/mlpiper/components/restful - we have whole orthogonal to mlpiper implementation of uwsgi/nginx. (it uses some utils/exceptions, should be changed to DRUM exception, or even its own). (It just can be a separate project).
> 
> We should move all the https://github.com/datarobot/mlpiper/tree/master/mlpiper-py/mlpiper/components/restful somewhere into DRUM and keep it a separate **uwsgi/nginx** implementation. For example to: https://github.com/datarobot/datarobot-user-models/tree/master/custom_model_runner/datarobot_drum/resource/components/restful.. Or even to the higher levels.
> 
> As for the RESTfulComponent https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful_component.py, we should move it next to UwsgiComponent(RESTfulComponent) https://github.com/datarobot/datarobot-user-models/blob/master/custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/uwsgi_serving.py (as you did). I would clean it and merge together.
> 
> In this way: we will use mlpiper's pipleline construction functionality and component's functionality(nginx) will be in DRUM
> 
> I don't think we should take some files from the standalone uwsgi/nginx implementation https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful_component.py https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful/uwsgi_broker.py https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful/uwsgi_ini_template.py and copy them directly into `uwsgi_component/` which defines `pipeline` logic, In this case we try use mlpiper's _**pipleline construction functionality**_ and in parallel use some uswgi/nginx functionality. Which is not good imho.
> 
> @zohar-mizrahi ?

UwsgiComponent already provides ability in DRUM to configure uwsgi/nginx configuration via mlpiper and we have no plans to change that functionality in scope of that story. We actually utilise that functionality to run DRUM with enabled threads to improve cpu utilization for proxy model use case. In our POC it was done via environment variable UWSGI_THREADS as that option was missed in [uwsgi_ini_template.py](https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful/uwsgi_ini_template.py) to allow set it in transparent way.
As result  in scope of that story we want just have all uwsgi options in UWSGI.ini file and not use environment variable like UWSGI_THREADS.

We copy just files which plan to improve a bit with minimum changes to avoid any change to existing functionality. As mlpiper repo is prohibited to update. Other case I would do those changes in mlpiper repo.
I can move those 3 files to any location  if needed  please advise but I would like to avoid any refactoring of those files to keep them close to original.

To remove/refactor full (or big part) of mlpiper looks like as separate tech debt story which require proper investigation and competency which I don't have at the moment. 

What about to have a meeting if we need to clarify any other items?

**Comment by yura-ant** [comment]
All tests pass expect formatting of mlpiper files. I would like to avoid formatting of those file via # pylint: disable-all  to keep them close to original files or I better format it?

**Review by eric-s-s — CHANGES_REQUESTED** [review]
based off of these changes, what is the actual aim here?  I see that https://github.com/datarobot/datarobot-user-models/pull/1091/commits/5fa4d6dd25462230abdb56b5f289c500c88ad7e2  is pulling this in and copying it.  

is the plan to then modify these specific pieces to change the eventual functionality?  It looks like that's the plan.  

## requested change

No matter then end use, this still needs to also pull in the unit tests for these components.  Especially if you plan on changing them, we'll want those test in place.

File: custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/restful_component.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+                          "`pip install uwsgi` or `pip install mlpiper[uwsgi]`"
+                raise MLPiperException(err_msg)
+
+        super(RESTfulComponent, self).__init__(
+            engine if engine else RestModelServingEngine("uwsgi-context")
+        )
+        self._dry_run = False
+        self._wsgi_broker = None
+        self._nginx_broker = None
+        self._wid = None
```
Comment (eric-s-s) [inline custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/restful_component.py:70]: what is `wid` ? and `wuuid` ?

**Comment by eric-s-s** [comment]
jira ticket: https://datarobot.atlassian.net/browse/CODEGEN-1918

project: https://datarobot.atlassian.net/browse/PBMP-6561
tech spec: https://docs.google.com/document/d/1yq-T7-pmjW49RsvbslNIl3BF_GRbYLti2McVmZgd_To/edit

**Comment by yakov-g** [comment]
@yura-ant 
We can have a meeting.
Sorry, but I still can not agree with the approach.
1. By copying those files we just introduce more imports from mlpiper while our goal is to stop using it.
2. The fact that you don't want to format added files and _keep them them close to the original_ introduces even more (mental?) binding between the drum and mlpiper, one always have to look back at the same files in mlpiper. It doesn't seem viable.

Another approach I could suggest.
0. Very very likely there will be a situation **_oh we need to copy one more file to DRUM_**
So
1. You **_copy_** all the uwsgi/nginx logic into DRUM maybe assuming it is WIP project.
2. Duplicate in DRUM RestfulComponent and UwsgiComponent or merge them together (UwsgiComponent2 + RestfulComponent2 ). It will be new (but truly just duplicated into DRUM)  component.
3. Add additional env var (similar to PRODUCTION) which make server use new UwsgiComponent2.
4. You change/improve/test/etc. 
5. Then we just switch RestfulComponent to this RestfulComponent2. Anyhow, as long as I'm aware, no one uses production mode. Maybe only @oleksandr-saienko ?

**Comment by yura-ant** [comment]
> based off of these changes, what is the actual aim here? I see that [5fa4d6d](https://github.com/datarobot/datarobot-user-models/commit/5fa4d6dd25462230abdb56b5f289c500c88ad7e2) is pulling this in and copying it.
> 
> is the plan to then modify these specific pieces to change the eventual functionality? It looks like that's the plan.
> 
> ## requested change
> No matter then end use, this still needs to also pull in the unit tests for these components. Especially if you plan on changing them, we'll want those test in place.

The aim is to allow us modify in separate PR [uwsgi_broker.py](https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful/uwsgi_broker.py) just to pass new threads option to uwsig.ini file via modify _generate_ini_file method and one constant [uwsgi_ini_template.py](https://github.com/datarobot/mlpiper/blob/master/mlpiper-py/mlpiper/components/restful/uwsgi_ini_template.py) that will make configuration of threads more transparent and in one place as today we can only set UWSGI_THREADS as environment variable when start drum to achieve the same functionality.

I was looking for such unit tests in mlpiper but unfortunately they are missed. I think they are missed as RESTfulComponent is abstract class where in drum we implement it and actually test it. The second file UwsgiBroker is kind of configuration of uwsgi server. I will add such unit test as I need to add here threads option in separate PR. 

Since I didn't change any functional logic in drum just change how we can modify desired files from mlpiper. As it  is prohibited to modify mlpiper repo.  So this PR looks like refactoring and we can merge it without new tests.

File: custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/restful_component.py (hunk 1 above)
Comment (yura-ant) [reply custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/restful_component.py:70]: wid is uwsgi.worker_id() 
wuuid  is str(uuid.uuid4()) just uniq id

**Comment by yura-ant** [comment]
> @yura-ant We can have a meeting. Sorry, but I still can not agree with the approach.
> 
> 1. By copying those files we just introduce more imports from mlpiper while our goal is to stop using it.
It is true only in term of explicit vs implicit dependency. At the end we still have the same imports


> 2. The fact that you don't want to format added files and _keep them them close to the original_ introduces even more (mental?) binding between the drum and mlpiper, one always have to look back at the same files in mlpiper. It doesn't seem viable.
> 
My main aim to keep it close to mlpiper to allow reviewers of PR spend less time to review changes. And yes it should simplify look back at same files in mlpiper for example to understood why we copy them and why they exist. Imagine  if mlpiper back to life then we can simply move configuration of thread option back to mlpiper and remove those files.

> Another approach I could suggest. 0. Very very likely there will be a situation **_oh we need to copy one more file to DRUM_** 
You right but every such situation is different user story with different resources and owners. 
For example somebody may have decide to not use mlpiper together with uwsgi at all and switch to own version of server which may run python programs or threads option will be deprecated at some point.
As Result 
Unfortunately with my limited experience with mlpiper and drum (this is my first task actually) I would avoid such big changes  as it could lead to mistakes and deadlines. 

> 
> 1. You **_copy_** all the uwsgi/nginx logic into DRUM, considering it new WIP logic.
> 2. Duplicate in DRUM RestfulComponent and UwsgiComponent or merge them together (UwsgiComponent2 + RestfulComponent2 ). It will be new (but truly just duplicated into DRUM)  component.
> 3. Add additional env var (similar to PRODUCTION) which make server use new UwsgiComponent2.
> 4. You change/improve/test/etc.
> 5. When it is stable. We just switch RestfulComponent to this RestfulComponent2. Anyhow, as long as I'm aware, no one uses production mode. Maybe only @oleksandr-saienko ?

**Comment by eric-s-s** [comment]
@yura-ant based on our discussions, should this PR be closed?

**Pull request closed without merging**